File: em-april-2026/dsa/sorting/foundation/radix_sort/main.py
BASE = 256


def radix_sort(arr: list[int]) -> list[int]:
    """
    Args:
     arr(list_int32)
    Returns:
     list_int32
    """
    n = len(arr)
    power = 1
    iter = 0
    while True:
        # print(f"At the start of iteration #{iter}, arr = {arr}")
        if all(num < power for num in arr):
            break  # if all the numbers in array are less than power, then there is nothing to do further
        buckets = [
            [] for i in range(BASE)
        ]  # total BASE buckets, one for each digit in 0-9.
        # iteration for power = BASE^i
        for num in arr:
            digit = (
                num // power
            ) % BASE  # extract the digit with value-multiplier = power.
            buckets[digit].append(num)
        # print(f"  At the iteration #{iter}, buckets = {buckets}")
        # if all the numbers are in the same bucket, then this iteration can be skipped
        if all(len(bucket) < n for bucket in buckets):
            # rearrange the numbers in arr based on the buckets.
            i = 0
            for bucket in buckets:
                arr[i : i + len(bucket)] = bucket
                i = i + len(bucket)
        # print(f"  At the end of iteration #{iter}, arr = {arr}")
        power = power * BASE
        ++iter
    # Write your code here.
    return arr


if __name__ == "__main__":
    arr = [(i + 1) for i in range(1000000)]
    # arr = [100000001, 0, 1000000000]
    orig_arr = arr.copy()
    radix_sort(arr)
    # print(f"radix_sort({orig_arr}) = {arr}")

File: em-april-2026/dsa/sorting/foundation/radix_sort/main.cpp
#include <algorithm>
#include <array>
#include <iostream>  // Added just in case you uncomment the prints
#include <numeric>
#include <vector>

/*
 * Args:
 * arr(std::vector<int>)
 * Returns:
 * std::vector<int>
 */
std::vector<int> radix_sort(std::vector<int> arr) {
    constexpr int BASE = 256;
    int iter = 0;

    // We embed the loop condition and power multiplication directly into the
    // for-loop. if all the numbers in array are less than power, then there is
    // nothing to do further
    for (long long power = 1;
         !std::all_of(arr.begin(), arr.end(), [=](int n) { return n < power; });
         power *= BASE) {
        // std::cout << "At the start of iteration #" << iter << "\n";

        // iteration for power = BASE^i
        // total BASE buckets, one for each digit in 0-255.
        std::array<std::vector<int>, BASE> buckets;
        for (int num : arr) {
            // extract the digit with value-multiplier = power.
            buckets[(num / power) % BASE].push_back(num);
        }

        // std::cout << "  At the iteration #" << iter << " buckets filled\n";

        // if all the numbers are in the same bucket, then this iteration can be
        // skipped
        if (std::none_of(buckets.begin(), buckets.end(), [&](const auto& b) {
                return b.size() == arr.size();
            })) {
            // rearrange the numbers in arr based on the buckets.
            auto it = arr.begin();
            for (auto& b : buckets) {
                // std::copy seamlessly acts like your python slice assignment
                it = std::copy(b.begin(), b.end(), it);
            }
        }

        // std::cout << "  At the end of iteration #" << iter << "\n";
        ++iter;
    }

    // Write your code here.
    return arr;
}

int main() {
    std::vector<int> arr(1000000);
    std::iota(arr.begin(), arr.end(), 1);

    // std::vector<int> arr = {100000001, 0, 1000000000};
    // std::vector<int> orig_arr = arr;

    arr = radix_sort(arr);

    return 0;
}
